src/primitives/gf32.rs

```diff
@@ -0,0 +1,481 @@
+    /// of the polynomial representation.
+    pub fn to_u8(self) -> u8 { self.0 }
+
+    fn _add(self, other: Fe32) -> Fe32 { Fe32(self.0 ^ other.0) }
+
+    // Subtraction is the same as addition in a char-2 field.
+    fn _sub(self, other: Fe32) -> Fe32 { self + other }
+
+    #[cfg_attr(all(test, mutate), mutate)]
+    fn _mul(self, other: Fe32) -> Fe32 {
+        if self.0 == 0 || other.0 == 0 {
+            Fe32(0)
+        } else {
+            let log1 = LOG[self.0 as usize];
+            let log2 = LOG[other.0 as usize];
+            Fe32(LOG_INV[((log1 + log2) % 31) as usize])
+        }
+    }
+
+    #[cfg_attr(all(test, mutate), mutate)]
+    fn _div(self, other: Fe32) -> Fe32 {
+        if self.0 == 0 {
+            Fe32(0)
+        } else if other.0 == 0 {
+            panic!("Attempt to divide {} by 0 in GF32", self);
+        } else {
+            let log1 = LOG[self.0 as usize];
+            let log2 = LOG[other.0 as usize];
+            Fe32(LOG_INV[((31 + log1 - log2) % 31) as usize])
+        }
+    }
+}
+
+impl fmt::Display for Fe32 {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result { fmt::Display::fmt(&self.to_char(), f) }
+}
+
+impl str::FromStr for Fe32 {
+    type Err = Error;
+    fn from_str(s: &str) -> Result<Fe32, Error> {
+        let mut chs = s.chars();
+        match (chs.next(), chs.next()) {
+            (Some(c), None) => Fe32::from_char(c),
+            (Some(_), Some(c)) => Err(Error::ExtraChar(c)),
+            (None, _) => Err(Error::InvalidLength(0)),
+        }
+    }
+}
+
+impl From<Fe32> for u8 {
+    fn from(v: Fe32) -> u8 { v.0 }
+}
+
+macro_rules! impl_try_from_upper_bounded {
+    ($($ty:ident)+) => {
+        $(
+            impl TryFrom<$ty> for Fe32 {
+                type Error = Error;
+
+                /// Tries to create a [`Fe32`] type from a source number type.
+                ///
+                /// # Errors
+                ///
+                /// Returns an error if `value` overflows an `Fe32`.
+                fn try_from(value: $ty) -> Result<Self, Self::Error> {
+                    let byte = u8::try_from(value)?;
+                    if byte > 31 {
+                        Err(Error::InvalidByte(byte))?;
+                    }
+                    Ok(Fe32(byte))
+                }
+            }
+        )+
+    }
+}
+macro_rules! impl_try_from_both_bounded {
+    ($($ty:ident)+) => {
+        $(
+            impl TryFrom<$ty> for Fe32 {
+                type Error = Error;
+
+                /// Tries to create a [`Fe32`] type from a signed source number type.
+                ///
+                /// # Errors
+                ///
+                /// Returns an error if `value` is outside of the range of an `Fe32`.
+                fn try_from(value: $ty) -> Result<Self, Self::Error> {
+                    if value.is_negative() {
+                        Err(Error::Negative)?;
+                    }
+                    let byte = u8::try_from(value)?;
```

Comment:
I think the negative check is entirely superfluous since `u8::try_from` will do it for us. So we can drop that, and drop the `Error::Negative` variant.